File: 链表/23.合并K个升序链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeKLists(lists []*ListNode) *ListNode {
	return mergeKList(lists, 0, len(lists))
}
func mergeKList(lists []*ListNode, i int, j int) *ListNode {
	m := j - i
	if m == 0 {
		return nil
	}
	if m == 1 {
		return lists[i]
	}
	left := mergeKList(lists, i, i+m/2)
	right := mergeKList(lists, i+m/2, j)
	return merge(left, right)
}

func merge(list1 *ListNode, list2 *ListNode) *ListNode {
	res := &ListNode{-1, nil}
	pre := res
	for list1 != nil && list2 != nil {
		if list1.Val <= list2.Val {
			pre.Next = list1
			list1 = list1.Next
		} else {
			pre.Next = list2
			list2 = list2.Next
		}
		pre = pre.Next
	}
	if list1 != nil {
		pre.Next = list1
	}
	if list2 != nil {
		pre.Next = list2
	}
	return res.Next
}

File: 链表/23.合并K个升序链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        return mergeKList(lists,0,lists.length);
    }

    ListNode mergeKList(ListNode[] lists,int i,int j){
        int m = j-i;
        if(m==0){
            return null;
        }
        if(m==1){
            return lists[i];
        }
        ListNode left = mergeKList(lists,i,i+m/2);
        ListNode right = mergeKList(lists,i+m/2,j);
        return merge(left,right);
    }

    private ListNode merge(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(); // 用哨兵节点简化代码逻辑
        ListNode cur = dummy; // cur 指向新链表的末尾
        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                cur.next = list1; // 把 list1 加到新链表中
                list1 = list1.next;
            } else { // 注：相等的情况加哪个节点都是可以的
                cur.next = list2; // 把 list2 加到新链表中
                list2 = list2.next;
            }
            cur = cur.next;
        }
        cur.next = list1 != null ? list1 : list2; // 拼接剩余链表
        return dummy.next;
    }

}